docs/customization.md

```diff
@@ -0,0 +1,167 @@
+> but allowing more wide-ranging customization.
+
+## Specifying a Functions Startup class
+
+When the Functions Framework is running normally (either in
+production or locally), Functions Startup classes are discovered via
+the `FunctionsStartupAttribute` assembly attribute. This attribute
+specifies a type to use as a Functions Startup class. The attribute
+can be specified multiple times, so you can separate your
+customization steps cleanly. In most cases the order in which
+Functions Startup classes are used may be irrelevant, but the
+attribute has an `Order` property to allow relative ordering where
+necessary. (This is most likely to be important when adding
+middleware to the request pipeline.)
+
+The Functions Startup class itself must be a subclass of
+`FunctionStartup`. This is an abstract class with virtual methods to
+override depending on what aspect you wish to customize. Later
+sections of this document demonstrate which methods are used to
+customize which aspect.
+
+Example:
+
+```csharp
+using Google.Cloud.Functions.Invoker;
+
+// Specify the Functions Startup class
+[assembly: FunctionsStartup(typeof(Example.Startup))]
+
+namespace Example
+{
+    public class Startup : FunctionsStartup
+    {
+        // Virtual methods in the base class are overridden
+        // here to perform customization.
+    }
+}
+```
+
+The attribute has to be in the same assembly as the function
+(because that's how it's discovered) but the Functions Startup class
+could be in another assembly. It has to be public, with a public
+parameterless constructor.
+
+When writing integration tests, it's often important to be able to
+swap out your "production" Functions Startup classes with "test"
+versions, for example to use fakes or mocks for services configured
+via dependency injection. The `FunctionTestServerBuilder` makes this
+easy by allowing Functions Startup instances to be specified while
+building the test server. See [the testing
+documentation](testing.md) for more details.
+
+## Customizing Dependency Injection using `ConfigureServices`
+
+The Functions Framework Invoker uses [ASP.NET Core dependency
+injection](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection)
+when constructing instances of the target function type to handle
+calls.
+
+See [SimpleDependencyInjection.Function](../examples/Google.Cloud.Functions.Examples.SimpleDependencyInjection/Function.cs)
+for an example of this, where the function constructor has an
+`ILogger<T>` parameter so that a logger can be injected.
+
+In more complex scenarios, users may wish to provide additional
+dependencies themselves through Functions Startup classes, by
+overriding the `ConfigureServices` method:
+
+```csharp
+public virtual void ConfigureServices(WebHostBuilderContext context, IServiceCollection services)
+```
+
+The dependencies are then applied by ASP.NET Core dependency
+injection in the normal way. The 
```

Comment:
```suggestion
injection in the normal way. The ???
```

Reply:
Hmm. That's odd. Not sure what I was trying to say there - have removed that sentence and brought the example into the same paragraph.